### DISCUSSION TITLE
Using WebGL extension `GL_OES_standard_derivatives`

### DISCUSSION THREAD

**Discussion by kahole** [visgl/deck.gl#8017] [Q&A] [answered]
Deck.gl version `8.9.22`

I'm having some trouble using the `GL_OES_standard_derivatives` functionality in shaders with Deck.gl.
Namely, when using `fwidth()` I'm getting an shader compilation error complaining that it doesn't exist. Which usually means you haven't enabled it or it's not supported in the browser.

I understand from the docs and reading online that `luma.gl` will, by default, include the directive for standard derivatives when possible.
I.e.
```
#extension GL_OES_standard_derivatives : enable
```
Is this an incorrect assumption for this case?

I've confirmed that my browser, latest version of Chrome, has no problem with the `GL_OES_standard_derivatives` extension outside of Deck.gl. 

By uncommenting `line 23` in the JS file of this [Codepen-instance](https://codepen.io/KaHole/pen/jOQpRWY) you'll find a minimal reproduction of the issue.

(https://codepen.io/KaHole/pen/jOQpRWY)

Any help getting derivatives working is greatly appreciated! And thank you to all the contributors and community members of Deck.gl, it's great!

---
JS and GLSL from the minimal example:
```
const { DeckGL, Layer, project32 } = deck;
const { Geometry, Model } = luma;

const vs = `
attribute vec3 positions;
attribute vec3 instancePositions;
attribute vec3 instancePositions64Low;

varying vec2 vPosition;

void main(void) {
  vec3 offsetCommon = positions * 10.0;
  vec3 positionCommon = project_position(instancePositions, instancePositions64Low);
  vPosition = positions.xy;

  gl_Position = project_common_position_to_clipspace(vec4(positionCommon + offsetCommon, 1.0));
}`;

const fs = `
varying vec2 vPosition;

void main(void) {
  float test_fwidth = fwidth(vPosition.x);
  gl_FragColor = vec4(vPosition.xy, 0.0, 1.0);
}`;

class TriangleLayer extends Layer {
  getShaders() {
    return {
      vs,
      fs,
      modules: [project32]
    };
  }
  
  initializeState() {
    const { gl } = this.context;

    this.getAttributeManager().addInstanced({
      instancePositions: {
        size: 3,
        type: gl.FLOAT,
        accessor: 'getPosition'
      },
    });
    
    this.setState({
      model: this.getModel(gl)
    });
  }
  
  getModel(gl) {
    if (!gl) {
      return null;
    }

    const positions = new Float32Array([-1, -1, -1, 1, 1, 1]);
    const geometry = new Geometry({
      drawMode: gl.TRIANGLES,
      vertexCount: 3,
      attributes: {
        positions: {size: 2, value: positions}
      }
    });
    return new Model(gl, {
      ...this.getShaders(),
      id: `${this.props.id  }-model`,
      isInstanced: true,
      geometry
    });
  }
}

TriangleLayer.layerName = "TriangleLayer";


const data = [
  { pos: [-80.0133149, 40.4554421] },
];

new DeckGL({
  container: document.getElementById("container"),
  initialViewState: {
    longitude: -80.0133149,
    latitude: 40.4554421,
    zoom: 4
  },
  controller: true,
  
  layers: [
    new TriangleLayer({
      id: 'test',
      data,
      getPosition: d => d.pos,
    })
  ]
});
```

**Comment by ibgreen** [comment, marked as answer]
Pay attention to the version of the WebGL context and your GLSL shaders. The extension is not available under WebGL 2, instead the functionality is built-in to GLSL 3.00. 

So... if it is WebGL2 you need to be using GLSL 3.00 to get this extension - add `#version 300 as` as the very first characters in your shader, luma.gl should recognize it and lift that declaration to the top of the assembled shader.

**Reply by kahole** [reply]
Thank you! `#version 300 es` works nicely.
So the recommended approach is to load different shaders based on whether the context is WebGL 1 or 2? (Or to simply prepend the version string based on version given compatability of the shader)

**Comment by ibgreen** [comment]
Yes, my recommendation would be to write all shaders in GLSL 3 (adding the `#version` directive like you did) and rely on luma.gl to transpile them back to GLSL 1 on WebGL 1 contexts.

**Reply by kahole** [reply]
Ah, cool! Thank you
